templates/explorer/explorer.py

```diff
@@ -0,0 +1,199 @@
+from js9 import j
+import os
+import time
+from zerorobot.template.base import TemplateBase
+from zerorobot.template.state import StateCheckError
+from zerorobot.service_collection import ServiceNotFoundError
+
+
+CONTAINER_TEMPLATE_UID = 'github.com/zero-os/0-templates/container/0.0.1'
+TFCHAIN_FLIST = 'https://hub.gig.tech/tfchain/ubuntu-16.04-tfchain-latest.flist'
+EXPLORER_FLIST = 'https://hub.gig.tech/tfchain/caddy-explorer-latest.flist'
+
+
+class Explorer(TemplateBase):
+    version = '0.0.1'
+    template_name = 'explorer'
+
+    def __init__(self, name=None, guid=None, data=None):
+        super().__init__(name=name, guid=guid, data=data)
+        self._tfchain_sal = None
+        self.recurring_action('_monitor', 30)  # every 30 seconds
+
+    def validate(self):
+        if not self.data.get('domain'):
+            raise ValueError("domain need to be specified")
+
+    @property
+    def _node_sal(self):
+        return j.clients.zero_os.sal.get_node(self.data['node'])
+
+    @property
+    def _container_sal(self):
+        return self._node_sal.containers.get(self._container_name)
+
+    @property
+    def _container_name(self):
+        return "container-%s" % self.guid
+
+    @property
+    def _explorer_sal(self):
+        kwargs = {
+            'name': self.name,
+            'container': self._container_sal,
+            'rpc_addr': '0.0.0.0:%s' % self.data['rpcPort'],
+            'api_addr': 'localhost:%s' % self.data['apiPort'],
+            'data_dir': '/mnt/data',
+            'domain': self.data['domain'],
+            'network': self.data.get('network', 'standard')
+        }
+        return j.clients.zero_os.sal.tfchain.explorer(**kwargs)
+
+    def _get_container(self):
+        sp = self._node_sal.storagepools.get('zos-cache')
+        try:
+            fs = sp.get(self.guid)
+        except ValueError:
+            fs = sp.create(self.guid)
+
+        # prepare persistant volume to mount into the container
+        node_fs = self._node_sal.client.filesystem
+        vol = os.path.join(fs.path, 'explorer')
```

Comment:
Trying to undesrtand why we use a fixed path 'explorer', and not something related to the running service. Will this not be blocking to run testnet and prodnet next to each other? Also, isn't this dangerous if we add other workloads beyond the tf chain on these boxes in the future?

Reply:
If you look line 55 I create a filesystem with the guid of the service, so there will be a different directory for every service. Using an extra explorer directory is to make it more easily readable in the mounts for debugging.